README.md
# Test Hibernate Connection with Gradle

The purpose of this project is to learn the basic of Hibernate using Gradle;
In this project will be used **Native Hibernate APIs**, **Annotation Mappings**
and **Criteria Queries**.

Please see [HIBERNATE Official Site](https://hibernate.org) for more informations.

## Requirements

### Base
* Java Used Version: 11
* [Gradle](https://gradle.org/) Used Version: 5.6.2
* [MySql](https://www.mysql.com) Used version: 8.0.15


### Dependencies

* [mysql-connector-java](https://bintray.com/bintray/jcenter/mysql%3Amysql-connector-java) Used Version: 8.0.17
* [hibernate-c3p0](https://bintray.com/bintray/jcenter/org.hibernate%3Ahibernate-c3p0/5.4.0.Final) Used Version: 5.4.0.Final
* [hibernate-core](https://bintray.com/bintray/jcenter/org.hibernate%3Ahibernate-core/5.4.0.Final) Used Version: 5.4.0.Final
* [junit](https://bintray.com/bintray/jcenter/junit%3Ajunit) Used Version: 4.12

## Configuration

rename the file  **src/main/resources/default_hibernate.cfg.xml** as **src/main/resources/hibernate.cfg.xml** and change the follow parameters with those of your MySql server
```
...
<property name="hibernate.connection.url">jdbc:mysql://localhost:3306/***DBNAME***?useSSL=false</property>
<property name="hibernate.connection.username">***USERNAME***</property>
<property name="hibernate.connection.password">***PASSWORD***</property>
...
```

### MySql Table

here the user Table to place into the Db
```
CREATE TABLE `users` (
  `username` varchar(10) NOT NULL,
  `password` varchar(250) NOT NULL,
  `email` varchar(50) NOT NULL,
  PRIMARY KEY (`username`)
)
```

## Changes 
Please see the [CHANGES](CHANGES.md) files for details

## Authors

* **[SirZiphon](https://github.com/SirZiphon)** - Initial work

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details


CHANGES.md
# Change Log

## V1.0.1
* minor fixes for launch javadoc generator task

## V1.0.0
* Initial Release

src/test/java/it/sirziphonprojects/UnitTest.java
/*
 * The MIT License
 *
 * Copyright 2019 SirZiphon <https://github.com/sirziphon>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package it.sirziphonprojects;

/**
 *
 * @author SirZiphon <https://github.com/sirziphon>
 * @version 1.0.0
 */
public interface UnitTest {
    
}


src/test/java/it/sirziphonprojects/dbconnection/entities/UsersTest.java
/*
 * The MIT License
 *
 * Copyright 2019 SirZiphon <https://github.com/sirziphon>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package it.sirziphonprojects.dbconnection.entities;

import it.sirziphonprojects.UnitTest;
import org.junit.Assert;
import org.junit.Test;
import org.junit.experimental.categories.Category;

/**
 *
 * @author SirZiphon <https://github.com/sirziphon>
 * @version 1.0.0
 */
@Category(UnitTest.class)
public class UsersTest {
    
    private final Users user;
    
    public UsersTest() {
        this.user = new Users("User1", "User1Pass", "User1email@mail.ma");
    }
    
    /**
     * Test of getUsername method, of class Users.
     */
    @Test
    public void testGetUsername() {
        System.out.println("getUsername");
        
        Assert.assertEquals("User1", this.user.getUsername());
    }

    /**
     * Test of setUsername method, of class Users.
     */
    @Test
    public void testSetUsername() {
        System.out.println("setUsername");
        
        this.user.setUsername("User2");
        Assert.assertEquals("User2", this.user.getUsername());
    }

    /**
     * Test of getPassword method, of class Users.
     */
    @Test
    public void testGetPassword() {
        System.out.println("getPassword");
        
        Assert.assertEquals("User1Pass", this.user.getPassword());
    }

    /**
     * Test of setPassword method, of class Users.
     */
    @Test
    public void testSetPassword() {
        System.out.println("setPassword");
        
        this.user.setPassword("User2Pass");
        Assert.assertEquals("User2Pass", this.user.getPassword());
    }

    /**
     * Test of getEmail method, of class Users.
     */
    @Test
    public void testGetEmail() {
        System.out.println("getEmail");
        
        Assert.assertEquals("User1email@mail.ma", this.user.getEmail());
    }

    /**
     * Test of setEmail method, of class Users.
     */
    @Test
    public void testSetEmail() {
        System.out.println("setEmail");
        
        this.user.setEmail("User2email@mail.ma");
        Assert.assertEquals("User2email@mail.ma", this.user.getEmail());
    }

    /**
     * Test of toString method, of class Users.
     */
    @Test
    public void testToString() {
        System.out.println("toString");

        String expResult = "User: (Username: User1, Password: User1Pass, Email: User1email@mail.ma)";
        Assert.assertEquals(expResult, this.user.toString());
    }
    
}


src/test/java/it/sirziphonprojects/dbconnection/models/UsersModelTest.java
/*
 * The MIT License
 *
 * Copyright 2019 SirZiphon <https://github.com/sirziphon>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package it.sirziphonprojects.dbconnection.models;

import it.sirziphonprojects.UnitTest;
import it.sirziphonprojects.dbconnection.entities.Users;
import it.sirziphonprojects.dbconnection.utils.HibernateUtil;
import java.util.List;
import org.junit.AfterClass;
import org.junit.Assert;
import org.junit.Test;
import static org.junit.Assert.*;
import org.junit.FixMethodOrder;
import org.junit.experimental.categories.Category;
import org.junit.runners.MethodSorters;

/**
 *
 * @author SirZiphon <https://github.com/sirziphon>
 * @version 1.0.0
 */
@Category(UnitTest.class)
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class UsersModelTest {

    private final UsersModel usersModel;
    private final Users userTest1;
    private final Users userTest2;
    
    public UsersModelTest() { 
        this.usersModel = UsersModel.getModel();
        this.userTest1 = new Users("test1", "test1pass", "test1@test1.test1");
        this.userTest2 = new Users("test2", "test2pass", "test2@test2.test2");
    }
    
    @AfterClass
    public static void end() {
        HibernateUtil.shutdown();
    }
    
    /**
     * Test of getModel method, of class UsersModel.
     */
    @Test
    public void a_testGetModel() {
        System.out.println("getModel");
        assertTrue(this.usersModel instanceof UsersModel);
    }

    /**
     * Test of getList method, of class UsersModel.
     */
    @Test
    public void d_testGetList() {
        System.out.println("getList");
        
        List<Users> list = this.usersModel.getList();
        
        assertTrue(list != null);
        assertTrue(list.size() >= 1);
    }

    /**
     * Test of getElement method, of class UsersModel.
     */
    @Test
    public void c_testGetElement() {
        System.out.println("getElement");

        Users foundElement = this.usersModel.getElement(this.userTest1.getUsername());

        if (foundElement != null) {
            assertTrue(this.userTest1.getUsername().equals(foundElement.getUsername()));
            assertTrue(this.userTest1.getPassword().equals(foundElement.getPassword()));
            assertTrue(this.userTest1.getEmail().equals(foundElement.getEmail()));
        } else {
            Assert.fail("no element found!");
        }
    }

    /**
     * Test of insertElement method, of class UsersModel.
     */
    @Test
    public void b_testInsertElement() {
        System.out.println("insertElement");

        boolean result1 = this.usersModel.insertElement(this.userTest1);
        boolean result2 = this.usersModel.insertElement(this.userTest2);

        assertTrue(result1);
        assertTrue(result2);
    }

    /**
     * Test of deleteElement method, of class UsersModel.
     */
    @Test
    public void e_testDeleteElement_Users() {
        System.out.println("deleteElement");
        
        assertTrue(this.usersModel.deleteElement(this.userTest1.getUsername()));
    }

    /**
     * Test of deleteElement method, of class UsersModel.
     */
    @Test
    public void f_testDeleteElement_String() {
        System.out.println("deleteElement");
        
        assertTrue(this.usersModel.deleteElement(this.userTest2));
    }

}


src/main/java/it/sirziphonprojects/dbconnection/entities/Users_.java
/*
 * The MIT License
 *
 * Copyright 2019 SirZiphon <https://github.com/sirziphon>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package it.sirziphonprojects.dbconnection.entities;

import javax.persistence.metamodel.SingularAttribute;
import javax.persistence.metamodel.StaticMetamodel;

/**
 *
 * @author SirZiphon {@literal <https://github.com/sirziphon>}
 * @version 1.0.1
 */
@StaticMetamodel(Users.class)
public class Users_ {
    public static volatile SingularAttribute<Users, String> username;
    public static volatile SingularAttribute<Users, String> password;
    public static volatile SingularAttribute<Users, String> email;
}


src/main/java/it/sirziphonprojects/dbconnection/models/BaseModel.java
/*
 * The MIT License
 *
 * Copyright 2019 SirZiphon <https://github.com/sirziphon>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package it.sirziphonprojects.dbconnection.models;

import it.sirziphonprojects.dbconnection.entities.BaseEntity;
import it.sirziphonprojects.dbconnection.utils.HibernateUtil;
import java.util.List;
import org.hibernate.SessionFactory;

/**
 * This class will be used as base for the implementation of all entities models
 * class
 * 
 * @author SirZiphon {@literal <https://github.com/sirziphon>}
 * @version 1.0.1
 * @param <T> the Entity class
 * @param <K> the type of the entity class primary key
 */
public abstract class BaseModel<T extends BaseEntity, K> {
    
    protected final SessionFactory sessionFactory;
    
    protected BaseModel() {
        this.sessionFactory = HibernateUtil.getSessionFactory();
    }
    
    /**
     * this method return the current list of entities
     * 
     * @return List of T entities
     */
    abstract public List<T> getList();
            
    /**
     * this method return the searched element; if no element match the passed 
     * key, <code>null</code> will be return
     * 
     * @param key the param to search the element 
     * @return the element searched or <code>null</code>
     */
    abstract public T getElement(K key);
    
    /**
     * this method insert the passed element into the db; return <code>true</code>
     * if the element is successfully inserted, <code>false</code> otherwise
     * 
     * @param element the element to insert
     * @return <code>true</code> for successfully insetion, <code>false</code> otherwise
     */
    abstract public boolean insertElement(T element);
    
    /**
     * this method delete the passed element from the db; return <code>true</code>
     * if the element is successfully deleted, <code>false</code> otherwise
     * 
     * @param element the element to delete
     * @return <code>true</code> for successfully deletion, <code>false</code> otherwise
     */
    abstract public boolean deleteElement(T element);
    
    /**
     * variation of the deleteElement(T element), in this case the element is delete
     * using only the primary key
     * 
     * @param key the primary key of the element to delete
     * @return  <code>true</code> for successfully deletion, <code>false</code> otherwise
     */
    abstract public boolean deleteElement(K key);
}


src/main/java/it/sirziphonprojects/dbconnection/models/UsersModel.java
/*
 * The MIT License
 *
 * Copyright 2019 SirZiphon <https://github.com/sirziphon>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package it.sirziphonprojects.dbconnection.models;

import it.sirziphonprojects.dbconnection.entities.Users;
import it.sirziphonprojects.dbconnection.entities.Users_;
import java.util.List;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;
import org.hibernate.Session;

/**
 *
 * Model class of the entity Users; this class provides all methods to manage the 
 * entity with the database (INSERT, SELECT, DELETE)
 * 
 * @author SirZiphon {@literal <https://github.com/sirziphon>}
 * @version 1.0.1
 */
public class UsersModel extends BaseModel<Users, String> {

    private static UsersModel singleton = null;

    private UsersModel() {
        super();
    }

    /**
     * funtion that return the UsersModel singleton instance; if no instance is
     * present, a new one is created
     *
     * @return the UsersModel instance
     */
    public static UsersModel getModel() {
        // check if the singleton is already istantiated
        if (UsersModel.singleton == null) {
            UsersModel.singleton = new UsersModel();
        }

        // return the singleton instance
        return UsersModel.singleton;
    }

    @Override
    public List<Users> getList() {
        List<Users> list = null;

        try (Session session = this.sessionFactory.getCurrentSession()) {
            session.beginTransaction();
            
            CriteriaQuery<Users> criteria = session.getCriteriaBuilder().createQuery(Users.class);
            criteria.from(Users.class);
            list = session.createQuery(criteria).getResultList();
            
            session.getTransaction().commit();
        } catch (Exception e) {
            System.err.println(e);
        }
        
        return list;
    }

    @Override
    public Users getElement(String key) {
        Users element = null;

        try (Session session = this.sessionFactory.getCurrentSession()) {
            session.beginTransaction();
            
            // create criteria Builder
            CriteriaBuilder builder = session.getCriteriaBuilder();
            
            CriteriaQuery<Users> criteria = builder.createQuery(Users.class);
            
            Root<Users> root = criteria.from(Users.class);
            
            criteria.select(root);
            
            criteria.where(builder.equal(root.get(Users_.username), key));
            
            List<Users> findedUsers = session.createQuery(criteria).getResultList();
            
            // check the result
            switch (findedUsers.size()) {
                case 0:
                    // no element found
                    break;
                    
                case 1: 
                    // element found
                    element = findedUsers.get(0);
                    break;
                    
                default: 
                    // more then one element found: WARNING! this case must be impossible because the passed parameter is a primary key
                    element = findedUsers.get(0);
                    System.err.println("WARNING: the method getElement of the Class UsersModel has found more then one tuple with the passed key: " + key);
            }
            
            session.getTransaction().commit();
        } catch (Exception e) {
            System.err.println(e);
        }
        
        return element;
    }

    @Override
    public boolean insertElement(Users element) {
        // check if the parameter is null
        if (element == null) 
            return false;   // no parameter; insertion failed
        
        // executing the db save procedure
        try (Session session = this.sessionFactory.getCurrentSession()) {
            session.beginTransaction();
            session.save(element);
            session.getTransaction().commit();
        } catch (Exception e) {
            System.err.println(e);
            return false;
        }
        
        // save successful
        return true;
    }

    @Override
    public boolean deleteElement(Users element) {
        // check if the parameter is null
        if (element == null) 
            return false;   // no parameter; deletion failed
        
        // executing the db deletion procedure
        try (Session session = this.sessionFactory.getCurrentSession()) {
            session.beginTransaction();
            session.delete(element);
            session.getTransaction().commit();
        } catch (Exception e) {
            System.err.println(e);
            return false;
        }
        
        // deletion successful
        return true;
    }

    @Override
    public boolean deleteElement(String key) {
        Users userKey;
        
        // check if the key is empty
        if (key == null || key.equals("")) 
            return false;   // no key; deletion failed
        else 
            userKey = this.getElement(key);
        
        // check if the user exist 
        if (userKey == null)
            return false; // no user; deletion failed
        
        // executing the db deletion procedure
        try (Session session = this.sessionFactory.getCurrentSession()) {
            session.beginTransaction();
            session.delete(userKey);
            session.getTransaction().commit();
        } catch (Exception e) {
            System.err.println(e);
            return false;
        }
        
        // deletion successful
        return true;
    }
}


src/main/java/it/sirziphonprojects/dbconnection/entities/BaseEntity.java
/*
 * The MIT License
 *
 * Copyright 2019 SirZiphon <https://github.com/sirziphon>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package it.sirziphonprojects.dbconnection.entities;

/**
 * This interface will be used for group all entity classes
 * 
 * @author SirZiphon {@literal <https://github.com/sirziphon>}
 * @version 1.0.1
 */
public interface BaseEntity {
    
}


src/main/java/it/sirziphonprojects/dbconnection/utils/HibernateUtil.java
/*
 * The MIT License
 *
 * Copyright 2019 SirZiphon <https://github.com/sirziphon>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package it.sirziphonprojects.dbconnection.utils;

import org.hibernate.SessionFactory;
import org.hibernate.boot.Metadata;
import org.hibernate.boot.MetadataSources;
import org.hibernate.boot.registry.StandardServiceRegistry;
import org.hibernate.boot.registry.StandardServiceRegistryBuilder;

/**
 * this class is used to bootstrap hibernate SessionFactory; 
 * 
 * @author SirZiphon {@literal <https://github.com/sirziphon>}
 * @version 1.0.1
 */
public class HibernateUtil {

    private static StandardServiceRegistry registry;
    private static SessionFactory sessionFactory;

    static {
        try {
            // Create registry
            registry = new StandardServiceRegistryBuilder().configure("hibernate.cfg.xml").build();
            // Create MetadataSources
            MetadataSources sources = new MetadataSources(registry);
            // Create Metadata
            Metadata metadata = sources.getMetadataBuilder().build();
            // Create SessionFactory
            sessionFactory = metadata.getSessionFactoryBuilder().build();
        } catch (Exception e) {
            System.err.println("Enitial SessionFactory creation failed" + e);
            if (registry != null) {
                StandardServiceRegistryBuilder.destroy(registry);
            }
        }
    }

    /**
     * this method return the created sessionFactory instance
     * 
     * @return sessionFactory instance
     */
    public static SessionFactory getSessionFactory() {
        return sessionFactory;
    }

    /**
     * this method close the db connection
     */
    public static void shutdown() {
        if (registry != null) {
            StandardServiceRegistryBuilder.destroy(registry);
        }
    }
}


src/main/java/it/sirziphonprojects/dbconnection/entities/Users.java
/*
 * The MIT License
 *
 * Copyright 2019 SirZiphon <https://github.com/sirziphon>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package it.sirziphonprojects.dbconnection.entities;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.Table;


/**
 *
 * This class represent the table <code>users</code> of the database
 * 
 * @author SirZiphon {@literal <https://github.com/sirziphon>}
 * @version 1.0.1
 */
@Entity
@Table(name="users")
public class Users implements BaseEntity {
   
    @Id
    @Column(name = "username", unique = true, nullable = false, length = 10)
    private String username;

    @Column(name = "password", unique = false, nullable = false, length = 250)
    private String password;
    
    @Column(name = "email", unique = false, nullable = false, length = 50)
    private String email;
    
    /**
     * constructor
     */
    public Users() { 
        this.username = null;
        this.password = null;
        this.email = null;
    }
    
    /**
     * Constructor with params
     * 
     * @param username
     * @param password
     * @param email 
     */
    public Users(String username, String password, String email) {
        this.username = username;
        this.password = password;
        this.email = email;
    }

    // getters and setters
    public String getUsername() { return this.username; }
    public void setUsername(String username) { this.username = username; }

    public String getPassword() { return this.password; }
    public void setPassword(String password) { this.password = password; }

    public String getEmail() { return this.email; }
    public void setEmail(String email) { this.email = email; }

    @Override
    public String toString() {
        return "User: (Username: " + this.username + ", Password: " + this.password + ", Email: " + this.email + ")";
    }    
}
